File: Olympiad/COCI/official/2009/contest3_solutions/cross.cpp
#include <iostream>
#include <vector>
using namespace std;

char ploca[9][10];

bool cross(int z) {
   bool bar_jednom = false;

   int red[9] = {0}, stupac[9] = {0}, blok[3][3] = {{0}};
   for ( int r=0; r<9; ++r ) {
      for ( int s=0; s<9; ++s ) {
         if ( ploca[r][s] == '0'+z ) {
            if ( red[r]++ > 0 ) throw 1;
            if ( stupac[s]++ > 0 ) throw 1;
            if ( blok[r/3][s/3]++ > 0 ) throw 1;
         }
      }
   }

   for ( int i=0; i<3; ++i ) {
      for ( int j=0; j<3; ++j ) {
         if ( blok[i][j] ) continue;

         vector<pair<int, int> > mog;
         for ( int r=0; r<3; ++r ) {
            for ( int s=0; s<3; ++s ) {
               if ( ploca[3*i+r][3*j+s] == '.' &&
                    !red[3*i+r] && !stupac[3*j+s] )
                  mog.push_back( make_pair(3*i+r, 3*j+s) );
            }
         }

         if ( mog.empty() ) throw 1;
         if ( mog.size() == 1 ) {
            int r = mog[0].first, s = mog[0].second;
            ploca[r][s] = '0'+z;
            red[r] = stupac[s] = blok[i][j] = 1;
            bar_jednom = true;
         }
      }
   }

   return bar_jednom;
}

int main() {
   for ( int i=0; i<9; ++i ) cin >> ploca[i];

   try {
      for ( ;; ) {
         bool gotovo = true;
         for ( int i=1; i<=9; ++i )
            gotovo &= !cross(i);
         if ( gotovo ) break;
      }
   } catch(...) {
      puts("ERROR");
      return 0;
   }

   for ( int i=0; i<9; ++i ) cout << ploca[i] << endl;

   return 0;
}

File: Olympiad/COCI/official/2009/contest3_solutions/cross.pas
program cross;

var
   i, j          : integer;
   z             : char;
   ploca         : array[0..8, 0..8] of char;
   ERROR, gotovo : boolean;

function cross(z : char):boolean;
var
   r, s, i, j          : integer;
   red, stupac         : array[0..8] of integer;
   blok                : array[0..2, 0..2] of integer;
   mog, gdje_r, gdje_s : integer;
begin
   fillchar(red, sizeof(red), 0);
   fillchar(stupac, sizeof(stupac), 0);
   fillchar(blok, sizeof(blok), 0);
   
   cross := false;
   for r:=0 to 8 do begin
      for s:=0 to 8 do begin
         if ploca[r, s] = z then begin
            red[r] := red[r]+1;
            stupac[s] := stupac[s]+1;
            blok[r div 3, s div 3] := blok[r div 3, s div 3]+1;
            if (red[r] > 1) or (stupac[s] > 1) or (blok[r div 3, s div 3] > 1) then
               ERROR := true;
         end;
      end;
   end;

   for i:=0 to 2 do begin
      for j:=0 to 2 do begin
         if blok[i, j] > 0 then continue;

         mog := 0;
         for r:=0 to 2 do begin
            for s:=0 to 2 do begin
               if (ploca[3*i+r, 3*j+s] = '.') and (red[3*i+r] = 0) and (stupac[3*j+s] = 0) then begin
                  mog := mog+1;
                  gdje_r := 3*i+r;
                  gdje_s := 3*j+s;
               end;
            end;
         end;

         if mog = 0 then begin
            ERROR := true;
         end;
         if mog = 1 then begin
            ploca[gdje_r, gdje_s] := z;
            red[gdje_r] := 1;
            stupac[gdje_s] := 1;
            blok[i, j] := 1;
            cross := true;
         end;
      end;
   end;
end;

begin
   for i:=0 to 8 do begin
      for j:=0 to 8 do
         read(ploca[i, j]);
      readln;
   end;

   repeat
      gotovo := true;
      for z:='1' to '9' do 
         gotovo := gotovo and not cross(z);
   until gotovo or ERROR;

   if ERROR then 
      writeln('ERROR')
   else begin
      for i:=0 to 8 do begin
         for j:=0 to 8 do
            write(ploca[i, j]);
         writeln;
      end;
   end;
end.
